### DISCUSSION TITLE
Property 'x' does not exist on type Ref

### DISCUSSION THREAD

**Discussion by mad-coder-365** [typegoose/typegoose#776] [General]
I am getting stuck with this basic error. Need help!

I have two classes

```ts
class Classroom {
  @prop()
  name: string;
}
```
```ts
class User {
  @prop()
  username: string;

  @prop(ref: () => Classroom)
  classId: Classroom;
}
```

When I map over array like

```ts
const users: User[] = userModel.find({}).populate('classId')  // classId is getting populated. This is working.

const usersInfo = users.map((user: User) => {
  return {
    id: user._id,
    className: user.classId.name
  }
})
```

I get two errors:
```
TS2339: Property '_id' does not exist on type 'User'.
```
```
TS2339: Property 'name' does not exist on type 'Ref<Classroom, ObjectId>'.   Property 'name' does not exist on type 'ObjectId'.
```

For the first error do I have to do something like:
```ts
public readonly _id: Types.ObjectId;
```

Best way to solve these errors? Thanks

**Comment by hasezoey** [comment]
>   classId: Classroom;

you are using the wrong type, it is `Ref<Class>`

> TS2339: Property '_id' does not exist on type 'User'.
> TS2339: Property 'name' does not exist on type 'Ref<Classroom, ObjectId>'.   Property 'name' does not exist on type 'ObjectId'.

you need to use typeguards, because a reference can be multiple types read [Reference other Classes: Population](https://typegoose.github.io/typegoose/docs/guides/advanced/reference-other-classes#population) or [API: typeguards: isDocument & isDocumentArray](https://typegoose.github.io/typegoose/docs/api/functions/typeguards/is-document)

PS: in case you are not on typegoose 10, you maybe need to shadow `_id` as a property in your class if you want access to it, like:
```ts
class YourClass {
  public _id?: mongoose.Types.ObjectId; // dont forget to NOT put a "@prop" decorator on this, otherwise you will be overwriting all defaults
}
```

**Reply by mad-coder-365** [reply]
Thanks rookie mistake 😁

**Reply by mad-coder-365** [reply]
I am on typegoose 10 but I think I referenced the class wrong. I will check.

**Comment by mad-coder-365** [comment]
@hasezoey I am on typegoose 10. I am still getting an error TS2339: Property '_id' does not exist on type User. If I shadow _id like you mentioned it works.

**Reply by hasezoey** [reply]
if you still get the error, then it is the problem of your re-casting of the type `user: User` at `users.map((user: User)`, just remove the type-recasting and you should get the proper type
Note: i didnt notice this recasting while doing my first answer

Edit: i also just noticed you also re-cast the type with `users: User[]` at `const users: User[]`, also dont do that, or use the proper type of `users: DocumentType<User>[] | null`

**Reply by mad-coder-365** [reply]
Makes sense, thanks :)
